### ISSUE TITLE
need accessibility

### ISSUE THREAD

**Issue by morethanreal** [PolymerElements/paper-input#5]

**Comment by morethanreal** [comment]
From @alice in https://github.com/PolymerElements/iron-input/issues/9:

Re invalid input:
I think the current mechanism of simply not allowing invalid input is confusing - note that does allow typing something other than numbers.
aria-invalid is intended to provide some hint that the value is invalid - https://www.w3.org/WAI/GL/wiki/Using_Aria-Invalid_to_Indicate_An_Error_Field#Example_2 shows a simple example of it in action, including some hint text shown when the input is invalid. Note the use of aria-describedby to attach the hint text: this is a good technique as it will be exposed as a help attribute, which tends to be spoken following a short delay after the name and value of the element.

[edit] - I see that it may use the invalid attribute already - a demo of that behaviour would be really useful here to see what the gaps are.

**Comment by lpalmaro** [comment]
Also, please note that the contrast of the hint text, labels, and password is not sufficient. Please darken that text so it is more visible.

**Comment by morethanreal** [comment]
What color should it be? It's the same color as the secondary menu item text, which is meeting the contrast requirements in this bug: https://github.com/PolymerElements/paper-item/issues/9

**Comment by lpalmaro** [comment]
Okay, if the secondary text is meeting the guidelines, then that is fine (although it's still hard to actually see for me, which is why I flagged it). The actual words "hint text" under "label", fourth from top, is super light though. Can we make that a darker shade at least?

**Comment by morethanreal** [comment]
You're right, that one is super light. I'll fix that one to match the label color.

**Comment by lpalmaro** [comment]
Thanks!

**Comment by morethanreal** [comment]
@blasten Can you set the placeholder to --secondary-text-color?

**Comment by blasten** [comment]
@morethanreal I checked the code and ran the example, but the placeholders & labels are already using `--secondary-text-color`  Is there one in particular you are looking at?

**Comment by lpalmaro** [comment]
It was "hint text" that was the lightest.

**Comment by morethanreal** [comment]
> Re invalid input:
> I think the current mechanism of simply not allowing invalid input is confusing - note that does allow typing something other than numbers.

What if we use `aria-live` to announce when the user tries to type a character, but the invalid input is not allowed?

I came across a problem case in the native `<input>`. If you have  `<input maxlength="5">` and try to type more than 5 characters, it doesn't add any new characters to the input, but also doesn't announce that no new characters have been entered.

**Comment by lpalmaro** [comment]
I like the idea of using aria-live and notifying the user that no
additional characters/invalid characters are allowed. Alice, what do you
think?

**Comment by alice** [comment]
My original comment referred to `<input type=number>` but I mis-formatted it :frowning: However I note that in Canary it now doesn't allow typing non-numbers. However, something like `<input type=tel pattern="[0-9]{3}"` is interesting in that it allows invalid input, doesn't display any visual errors but does set `invalid` in the accessibility tree. (In FF it has a red border when you tab away as well.)

I think we should try and match the spoken UI as closely as possible to the visible UI. Still thinking about what exactly that should look like - heading in to office now, will ping back one I'm in.

**Comment by alice** [comment]
I notice that `paper-input` shows an error rather than simply disallowing input, which was my original issue. It also sets its `invalid` state correctly. My only suggestion at this point would be to associate the `paper-input-error` with the internal `input` field using `aria-describedby` (this exposes it as "help text" rather than a label - the label should be whatever the form label is in the context).

Speaking of labels, I note in the demo that the nearby "only type letters" label isn't associated with the input. Is this a bug in the demo?

**Comment by morethanreal** [comment]
I added the `aria-labelledby` attribute in 50081f3406a0facc58027e014e25a49eede2b379 last night, maybe update?

**Comment by morethanreal** [comment]
The developer has a few options wrt validation in the input:

1. manual validation only - the input does not show an error until you call `validate()`, at which point it shows the error (second demo in "validation" section)
2. auto validation - the error state updates as the user types. You can type in invalid input. (first demo in "validation" section)
  * Right now it announces with a `role=alert`
3. prevent invalid input - the user cannot type in invalid input. The error state is not shown. (no demo currently)

**Comment by alice** [comment]
I see the code in there in my local version, but I don't see `aria-labelledby` behaviour in the demo.

**Comment by morethanreal** [comment]
Closed by #45.

**Comment by lpalmaro** [comment]
Awesome - Monica, let me know if I can test the most recent demo of this
when we go over the gold demos.
